Add support for adding widgets

westonruter opened the pull request:
It would be great if the great new widgets UI you're working on could be embedded within in a lightbox window that appears over the window when using the customizer (just like when you open the media manager). Widgets could then be dragged into the relevant customizer sections in the customizer panel (each sidebar could have its own section), and the preview window could then update to show the user what the site looks like with that widget added. 
- [x] Expand the widget control when it is first added to the sidebar.
- [x] Prepend the added widget to the existing widgets added to the sidebar. (Need to prepend the `widgets_sidebars` setting.)
- [x] Is WordPress keeping track of the multi_number outside of the widget form?
- [x] Don't invoke the widget save option when first adding a widget (to avoid error messages).
- [x] When removing a single widget, make option to add the widget enabled again.
- [x] Fix restoring of single widgets (has initial previewing state, and instance override does not work since control looks at `$_POST`)
- [x] When initially adding a widget, make sure it will appear in the preview even before saving.
- [x] When adding a single widget which previously existed (and is now in inactive sidebar), the old instance should remain.
- [x] Refactor all of this to eliminate PHP 5.3 dependencies
- [x] Remove pre-preview filters once `customize_register` happens
- [x] Refactor option transactions into separate class.
- [x] What is wrong with adding the RSS Feed widget?
- [x] Adding or reusing nonces for security? (e.g. in `preview_new_widgets`)

westonruter commented:
See also the following customizer concept from the Widgets UI Refresh, featured in the [Aug 26, 2013 meeting summary](http://make.wordpress.org/ui/2013/08/28/widgets-widgets-widgets-meeting-aug-26-2013/):

![image](https://f.cloud.github.com/assets/134745/1162927/0cd01716-202d-11e3-8203-235bbad354a4.png)

westonruter commented:
See also my [concerns](http://make.wordpress.org/ui/2013/08/28/widgets-widgets-widgets-meeting-aug-26-2013/#comment-23903) about the above concept, specifically the "Select a template to add widgets to" part.

westonruter commented:
I just realized a compact UI for adding widgets the Widget Customizer: as part of the (hidden) control for storing the widget orders, we could also include a drop-down of all registered widgets along with a button to add. Clicking **add** would insert a new widget form control at the top of the list, after which it could then be dragged to its desired location (see #1).

shaunandrews commented:
Another idea for adding widgets by hovering in empty spaces in an area:

![screen shot 2013-09-27 at 12 03 13 pm](https://f.cloud.github.com/assets/191598/1227281/68f1b4a6-278e-11e3-9b04-15d24b6ebe2c.png)

bobbravo2 commented:
@westonruter I was initially thinking this made sense as an additional control - however - the SRP dictates that there be two separate controls: one for adding new widgets, and another to control widget preview order.

I'll give you some feedback on my thoughts RE: separate control for widget order on #14

westonruter commented:
@bobbravo2 sorry, what does “SRP” mean?

shaunandrews commented:
Related to @westonruter recent comment on #16, here's a minor tweak that would work better with horizontally oriented widgets:

![screen shot 2013-09-28 at 11 52 55 am](https://f.cloud.github.com/assets/191598/1231456/13d36096-2856-11e3-8ab1-c55a8fe51cf3.png)

westonruter commented:
@shaunandrews so for horizontally-oriented widgets, would this **ADD WIDET** divider/button be rotated 90°?

bobbravo2 commented:
@westonruter sorry... I'm an acronym geek..: SRP Single Responsibility Principal.

westonruter commented:
@bobbravo2 thanks :smile: So regarding SRP, I think what I worked on over the weekend makes sense. The new control is "sidebar widgets" and it is for the setting which is just an array of the widget IDs contained in the sidebar. So since the setting itself handles both the ordering of the widgets, and which widgets are contained in the sidebar, it seems the customizer control can also logically manage both the ordering and the adding/removing of widgets from sidebars. 

I think it worked out quite cleanly, if you look at the changes on #1 and then how easy it was to add dragging between sidebars in #24.

westonruter committed:
Export available widgets to JS and add widget select

westonruter commented:
I'm doing work in the [`issue-3-addition`](https://github.com/x-team/wp-widget-customizer/tree/issue-3-addition) branch. Here's the initial UI for where a widget would be added to a sidebar:

![customize__wordpress_develop___just_another_wordpress_site](https://f.cloud.github.com/assets/134745/1363689/37014ddc-3830-11e3-8c18-5c196f13f1a5.png)

westonruter committed:
Populate available widgets select with Backbone collection

westonruter committed:
Refactor building of setting args into helper method

westonruter committed:
Register non-active widget settings as well

westonruter committed:
Only apply jshint to staged JS files

westonruter committed:
Add ajax groundwork for handing newly added widgets

westonruter commented:
Screenshot with the select dropdown opened:

![image](https://f.cloud.github.com/assets/134745/1376060/acac8b50-3a8d-11e3-96eb-1a3e0a04f766.png)

westonruter committed:
Move setupDeleteControl from sidebar_widgets into widget_form

Closes #41

westonruter committed:
Add support for single (non-multi) widgets

westonruter committed:
Prepare widget control for instantiation

westonruter committed:
Use filter_input() and do widget option update sanity check

westonruter committed:
Export default transport with each availavle widget

westonruter committed:
Add new widget setting and control to section

Setting is not yet hooked up with preview

westonruter committed:
Filter preview so incoming widget additions will be recognized

westonruter committed:
Refactor logic for deleting widgets to rely on model

westonruter committed:
Make widget addition, deletion, and reordering all model-driven

Allow old widgets to be resurrected, e.g. via Settings Revisions
Ensure widgets added in customizer persist

westonruter committed:
Fold standalone updatePreview method into sortable update event

westonruter commented:
I've reached a beta state for this widget addition feature.

westonruter committed:
Make old single widget available again after removal

westonruter committed:
Reorganize sidebar_widgets model setup

westonruter committed:
Fix failure for new widget additions to trigger preview refresh

westonruter committed:
Prevent old single widgets from not obeying instance override

westonruter committed:
Move option transacrions into separate class

westonruter committed:
Fix up handling of RSS widget which fails to template itself

Remove dead code

bobbravo2 commented:
@westonruter love the drop down for adding new widgets - that's what I was hoping for. What do you think about adding the Chosen jQuery plugin to enhance it with search (search a widget's description / title / possible alternate keywords for new users)?

Chosen Demo
http://harvesthq.github.io/chosen/

westonruter committed:
Eliminate retrieve_widgets() by managing wp_inactive_widgets directly

westonruter committed:
Elimiante closures for PHP 5.2 compat

westonruter committed:
Fix typo which broke widget additions

westonruter committed:
Fix dragging between sidebars

westonruter committed:
Keep form_control sidebar_id accurate when dragged to another sidebar

westonruter commented:
@bobbravo2 yes, I'd love to add Chosen or [Select2](http://ivaynberg.github.io/select2/) to improve the usability of the dropdown. From the Chosen demo, I assume the Standard Select or the Standard Select with optgroup would be the field we'd be interested in? I guess too we'll want to let the autocomplete match not only the title of the widget, but its description as well?

westonruter committed:
Fix trashing widgets with corrected jQuery.contains params

westonruter committed:
Remove RSS Feed widget fixups

The retrieve_widgets() function may have been causing the problem.

westonruter committed:
Add nonce checks for prepreview setup

westonruter commented:
Well, that was an epic effort. So happy with how it turned out! :satisfied:

westonruter commented:
To see the behavior for old-style single (non-multi) widgets, try this plugin which adds a test one: https://gist.github.com/westonruter/7141599

jtsternberg commented:
+1 for Select2
